src/charm.py

```diff
@@ -0,0 +1,378 @@
+
+    def _write_config_file(
+        self,
+        content: str,
+    ) -> None:
+        """Writes config file to workload.
+
+        Args:
+            content (str): Config file content.
+        """
+        self._container.push(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE}", source=content)
+        logger.info("Pushed: %s to workload.", CONFIG_FILE)
+
+    def _write_ue_config_file(self) -> None:
+        """Writes UE config file to workload."""
+        with open(f"src/{UEROUTING_CONFIG_FILE}", "r") as f:
+            content = f.read()
+
+        self._container.push(path=f"{BASE_CONFIG_PATH}/{UEROUTING_CONFIG_FILE}", source=content)
+        logger.info("Pushed %s config file to workload", UEROUTING_CONFIG_FILE)
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns whether a given Juju relation was crated.
+
+        Args:
+            relation_name (str): Relation name
+
+        Returns:
+            bool: Whether the relation was created.
+        """
+        return bool(self.model.get_relation(relation_name))
+
+    @property
+    def _config_file_is_written(self) -> bool:
+        """Returns whether the config file was written to the workload container.
+
+        Returns:
+            bool: Whether the config file was written.
+        """
+        return bool(self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE}"))
+
+    def _render_config_file(
+        self,
+        *,
+        default_database_name: str,
+        default_database_url: str,
+        smf_database_name: str,
+        smf_url: str,
+        smf_sbi_port: int,
+        nrf_url: str,
+        pod_ip: str,
+    ) -> str:
+        """Renders the config file content.
+
+        Args:
+            default_database_name (str): Database name.
+            default_database_url (str): Database URL.
+            smf_database_name (str): SMF database name.
+            smf_url (str): SMF URL.
+            smf_sbi_port (int): SMF SBI port.
+            nrf_url (str): NRF URL.
+            pod_ip (IPv4Address): Pod IP address.
+
+        Returns:
+            str: Config file content.
+        """
+        jinja2_env = Environment(loader=FileSystemLoader("src/templates"))
+        template = jinja2_env.get_template("smfcfg.yaml.j2")
+        return template.render(
+            default_database_name=default_database_name,
+            default_database_url=default_database_url,
+            smf_database_name=smf_database_name,
+            smf_url=smf_url,
+            smf_sbi_port=smf_sbi_port,
+            nrf_url=nrf_url,
+            pod_ip=pod_ip,
+        )
+
+    def _config_file_content_matches(self, content: str) -> bool:
+        """Returns whether the config file content matches the provided content.
+
+        Returns:
+            bool: Whether the config file content matches
+        """
+        existing_content = self._container.pull(path=f"{BASE_CONFIG_PATH}/{CONFIG_FILE}")
+        return existing_content == content
```

Comment:
Have you tested this? Are you sure this works? The output of `pull` is a StringIO object and if you want to compare strings you'll have to `.read()` it.

Reply:
Fixed. I had tested it but I had a bug in the test, I was patching the pull return value to be a string, and it's a `StringIO` object as you've point out. Good catch.

Reply:
I've set the itest relying on mongo to xfail. I can create an issue. The crash is related with some lib used in the `tls-certificates-interface`
```logs
unit-sdcore-smf-operator-0: 2023-05-19 12:44:33 DEBUG juju.worker.uniter.storage next hook op for storage-config-0: {Kind:2 Life:alive Attached:true Location:/var/lib/juju/storage/config/0}
unit-sdcore-smf-operator-0: 2023-05-19 12:44:33 DEBUG juju.worker.uniter no operations in progress; waiting for changes
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install Traceback (most recent call last):
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install   File "./src/charm.py", line 34, in <module>
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install     from charms.mongodb.v0.mongodb_tls import MongoDBTLS
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install   File "/var/lib/juju/agents/unit-mongodb-k8s-0/charm/lib/charms/mongodb/v0/mongodb_tls.py", line 16, in <module>
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install     from charms.tls_certificates_interface.v1.tls_certificates import (
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install   File "/var/lib/juju/agents/unit-mongodb-k8s-0/charm/lib/charms/tls_certificates_interface/v1/tls_certificates.py", line 257, in <module>
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install     from cryptography import x509
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install   File "/var/lib/juju/agents/unit-mongodb-k8s-0/charm/venv/cryptography/x509/__init__.py", line 6, in <module>
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install     from cryptography.x509 import certificate_transparency
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install   File "/var/lib/juju/agents/unit-mongodb-k8s-0/charm/venv/cryptography/x509/certificate_transparency.py", line 10, in <module>
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install     from cryptography.hazmat.bindings._rust import x509 as rust_x509
unit-mongodb-k8s-0: 2023-05-19 12:44:34 WARNING unit.mongodb-k8s/0.install ImportError: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /var/lib/juju/agents/unit-mongodb-k8s-0/charm/venv/cryptography/hazmat/bindings/_rust.abi3.so)
unit-mongodb-k8s-0: 2023-05-19 12:44:34 ERROR juju.worker.uniter.operation hook "install" (via hook dispatching script: dispatch) failed: exit status 1
unit-mongodb-k8s-0: 2023-05-19 12:44:34 DEBUG juju.machinelock created rotating log file "/var/log/machine-lock.log" with max size 10 MB and max backups 5
unit-mongodb-k8s-0: 2023-05-19 12:44:34 DEBUG juju.machinelock machine lock "machine-lock" released for mongodb-k8s/0 uniter (run install hook)
unit-mongodb-k8s-0: 2023-05-19 12:44:34 DEBUG juju.worker.uniter.operation lock released for mongodb-k8s/0
unit-mongodb-k8s-0: 2023-05-19 12:44:34 INFO juju.worker.uniter awaiting error resolution for "install" hook
unit-mongodb-k8s-0: 2023-05-19 12:44:34 DEBUG juju.worker.uniter [AGENT-STATUS] error: hook failed: "install"
unit-mongodb-k8s-0: 2023-05-19 12:44:34 DEBUG juju.worker.uniter.remotestate storage attachment change for mongodb-k8s/0: {storage-db-1 {2 alive true /data/db}}
unit-mongodb-k8s-0: 2023-05-19 12:44:34 INFO juju.worker.uniter awaiting error resolution for "install" hook
model-72ff20dc-7a57-45fb-8871-87ea73330d60: 2023-05-19 12:44:34 DEBUG juju.worker.caasadmission received admission request for sdcore-smf-operator-0 of /v1, Kind=Pod in namespace test-integration-gxwm
controller-0: 2023-05-19 12:44:34 DEBUG juju.worker.caasapplicationprovisioner.sdcore-smf-operator updating application "sdcore-smf-operator" trust to false
model-72ff20dc-7a57-45fb-8871-87ea73330d60: 2023-05-19 12:44:34 DEBUG juju.worker.caasadmission received admission request for sdcore-smf-operator of rbac.authorization.k8s.io/v1, Kind=Role in namespace test-integration-gxwm
model-72ff20dc-7a57-45fb-8871-87ea73330d60: 2023-05-19 12:44:34 DEBUG juju.worker.caasadmission received admission request for test-integration-gxwm-sdcore-smf-operator of rbac.authorization.k8s.io/v1, Kind=ClusterRole in namespace 
unit-sdcore-smf-operator-0: 2023-05-19 12:44:36 DEBUG juju.worker.uniter.remotestate storage attachment change for sdcore-smf-operator/0: {storage-config-0 {2 alive true /var/lib/juju/storage/config/0}}
unit-sdcore-smf-operator-0: 2023-05-19 12:44:36 DEBUG juju.worker.uniter.storage next hook op for storage-config-0: {Kind:2 Life:alive Attached:true Location:/var/lib/juju/storage/config/0}
unit-sdcore-smf-operator-0: 2023-05-19 12:44:36 DEBUG juju.worker.uniter no operations in progress; waiting for changes
```